### PR TITLE
Add support for defining C variadic functions

### LINKED ISSUE #515: support exporting a c-style var args function
Generally we don't want zig programmers to use c-style var args functions. But sometimes you have to interface with C code.

Here are some use cases:

 * implementing libc in zig (#514)
 * handling correctly this code from MSVC stdio.h:

```c
_Check_return_opt_
_CRT_STDIO_INLINE int __CRTDECL printf(
    _In_z_ _Printf_format_string_ char const* const _Format,
    ...)
#if defined _NO_CRT_STDIO_INLINE
;
#else
{
    int _Result;
    va_list _ArgList;
    __crt_va_start(_ArgList, _Format);
    _Result = _vfprintf_l(stdout, _Format, NULL, _ArgList);
    __crt_va_end(_ArgList);
    return _Result;
}
#endif
```

LLVM docs on how to do it:
http://llvm.org/docs/LangRef.html#variable-argument-handling-intrinsics

### PR DESCRIPTION
**Pull request by tadeokondrak**
Closes #515 (wow, low number!).

### CODE REVIEW

**Comment by tadeokondrak** [comment]
Hmm, I've realized now that the ABI won't match for passing a VaList as a C function parameter, you'd need to pass it as a pointer.
So a signature in C can be
```c
int vprintf(const char *format, va_list ap);
```
but the equivalent in Zig would be
```zig
extern "c" fn vprintf(format: [*:0]const u8, ap: *builtin.VaList) c_int;
```
Not sure how to solve this, since it seems designed around C's array decay semantics.

**Comment by andrewrk** [comment]
Wow this is a big feature! I'm happy to work with you on getting this landed, but I want to run another idea by you, which would be to ignore this for stage1 and only implement it in stage2. What do you think about that?

**Comment by andrewrk** [comment]
> Not sure how to solve this, since it seems designed around C's array decay semantics.

I'm not sure I fully understand the problem here, but just in case you weren't aware: the way that structs are passed is different with C calling convention functions than auto. So we should have no problem matching the C ABI in C calling convention functions here.

**Comment by tadeokondrak** [comment]
> Wow this is a big feature! I'm happy to work with you on getting this landed, but I want to run another idea by you, which would be to ignore this for stage1 and only implement it in stage2. What do you think about that? (Keep in mind this is planned now: #5246 (comment)) So with any improvements we make to stage1, really the only code that will survive long term is additional test coverage.

I think it's useful to implement this in stage1 first because
    - It's useful, there are a bunch of C libraries that use a callback that takes a `va_list` for logging.
    - It's easy, since LLVM handles all the code generation.
    - It helps to fully decide the API and see the edge cases before it's implemented in stage2.

Also I don't think it's very big of an implementation, a lot of the code is just boilerplate adding 8 new IR instructions.

> I'm not sure I fully understand the problem here, but just in case you weren't aware: the way that structs are passed is different with C calling convention functions than auto. So we should have no problem matching the C ABI in C calling convention functions here.

The problem is that on some targets, like x86_64 System V, `va_list` is defined like this:
```c
struct va_list_tag {
    unsigned int gp_offset;
    unsigned int fp_offset;
    void *overflow_arg_area;
    void *reg_save_area;
};

typedef struct va_list_tag va_list[1];
```
so that when it's defined on the stack with
```c
va_list ap;
```
it has the full 4 fields, but when it's passed as a parameter like
```c
int vprintf(const char *format, va_list ap);
```
it's passed as a single pointer.

This alone wouldn't be a big problem, but some targets (like Windows) define `va_list` as
```c
typedef char *va_list;
```
so it has the same representation when on the stack and when passed as a pointer.

The first solution I thought of was have something like to
```zig
pub const VaList =
    if (builtin.os.tag == .windows)
        *c_char
    else
        struct { ... };

pub const VaListRef =
    if (builtin.os.tag == .windows)
        VaList
    else
        *VaList;
```
where every function that takes a VaList by parameter would use VaListRef.

But this is still a problem because of whether or not you need to use & to pass a VaList now depends on the target,
so it's essentially impossible to write portable code with it.

Rust's unstable C variadic support seems to rely on some tricks related to overloading the dereference operator.
I'm not aware of any other languages that have support for interacting with `va_list`.

File: test/compile_errors.zig [hunk 1]
```diff
@@ -7419,7 +7435,7 @@ pub fn addCases(cases: *tests.CompileErrorContext) void {
     });
 
     cases.add( // fixed bug #2032
-        "compile diagnostic string for top level decl type",
+    "compile diagnostic string for top level decl type",
```
Comment (daurnimator) [inline test/compile_errors.zig:7438]: what's going on with the indentation here?

**Review by daurnimator — COMMENTED** [review]
Great to see someone tackle this!

Misc thoughts that don't necessarily need to be addressed:
  - perhaps `vaEnd` and `vaArg` should be primarily exposed as methods on `VaList`? `vaStart` could be a namespaced function there too. In future these could be implemented with inline assembly or https://github.com/ziglang/zig/issues/4466
  - this would imply that `VaList` would need to always be a struct (e.g. on windows could it be `extern struct { ptr: *c_char }`)
  - should we remove the `__` prefix from the VaList fields on some architectures?

**Comment by andrewrk** [comment]
> But this is still a problem because of whether or not you need to use & to pass a VaList now depends on the target,
> so it's essentially impossible to write portable code with it.

I'm still not following. Where would we need to put the `&`? Can you show some example code demonstrating the problem?

**Comment by tadeokondrak** [comment]
@daurnimator 
> perhaps vaEnd and vaArg should be primarily exposed as methods on VaList? vaStart could be a namespaced function there too. In future these could be implemented with inline assembly or #4466

I think these are used little enough that the confusion of these being methods rather than builtins would outweigh the convenience. 
Also inline assembly isn't enough to implement these, see <https://git.musl-libc.org/cgit/musl/commit/?id=ffaaa6d230512f3a7f3d040b943517728f3dc3cf>.

> this would imply that VaList would need to always be a struct (e.g. on windows could it be extern struct { ptr: *c_char })

I don't think that necessarily has the same ABI.

> should we remove the __ prefix from the VaList fields on some architectures?

I think it's better not to, just for translate-c code which might theoretically use the fields directly.

@andrewrk 

> I'm still not following. Where would we need to put the &? Can you show some example code demonstrating the problem?

```zig
const builtin = @import("builtin");

// If VaList is defined as a struct (like x86_64 System V)
// Note that @vaStart and @vaCopy have to return the full struct

fn vadd(n: c_int, ap: *builtin.VaList) callconv(.C) c_int {
    var i: c_int = 0;
    var res: c_int = 0;
    while (i < n) : (i += 1) {
        res += @vaArg(ap, c_int);
    }
    return res;
}

fn add(n: c_int, ...) callconv(.C) c_int {
    var ap = @vaStart();
    defer @vaEnd(&ap);
    return vadd(n, &ap);
}

// If VaList is defined as a pointer (like Windows)

fn vadd(n: c_int, ap: builtin.VaList) callconv(.C) c_int {
    var i: c_int = 0;
    var res: c_int = 0;
    while (i < n) : (i += 1) {
        res += @vaArg(ap, c_int);
    }
    return res;
}

fn add(n: c_int, ...) callconv(.C) c_int {
    var ap = @vaStart();
    defer @vaEnd(ap);
    return vadd(n, ap);
}
```

File: test/compile_errors.zig (hunk 1 above)
Comment (tadeokondrak) [reply test/compile_errors.zig:7438]: I think that's related to the recent zig fmt changes. @LakeByTheWoods is this a bug?

**Comment by andrewrk** [comment]
Thanks.

I would expect it to look like the latter code in both cases. Why wouldn't that work for x86_64 System V? These builtins are language primitives; not functions. They do not have to obey the rule of not modifying parameters.

**Comment by tadeokondrak** [comment]
> I would expect it to look like the latter code in both cases. Why wouldn't that work for x86_64 System V? These builtins are language primitives; not functions. They do not have to obey the rule of not modifying parameters.

True, but passing it to regular functions that take a VaList also has this problem (see the vadd call).

Though, a possible solution is to have `@vaStart`/`@vaCopy` on struct platforms allocate the full struct on the stack but return a pointer to it. This breaks storing the full VaList struct anywhere else, like in another stack-allocated struct or a global, but I doubt that's ever actually needed.

**Comment by andrewrk** [comment]
> True, but passing it to regular functions that take a VaList also has this problem (see the vadd call).

Sorry, I'm not trying to play dumb - What's the problem here? An `extern struct` will have the correct ABI here, right?

**Comment by tadeokondrak** [comment]
In C (on struct platforms), a `va_list` is defined as a `va_list_tag[1]`, which means it has a different size on the stack or in a struct than passed as a parameter.

It's similar to this:
```c
void foo(char a[4]);

void bar(void) {
    char a[4];
    foo(a);
}
```
In the function definition, `char a[4]` decays to being the same as `char *a`, and when passing it as a parameter, `a` decays to being just a pointer to the array, rather than passing it by value.

**Comment by andrewrk** [comment]
You can just make it a `[1]extern struct{...}` and get the exact same behavior. That's how C calling convention works in Zig.

**Comment by tadeokondrak** [comment]
Ah, so
```zig
export fn foo(a: [4]u8) callconv(.C) void {
    @intToPtr([*]u8, @ptrToInt(&a))[0] = 'a';
    @intToPtr([*]u8, @ptrToInt(&a))[1] = 'b';
    @intToPtr([*]u8, @ptrToInt(&a))[2] = 'c';
    @intToPtr([*]u8, @ptrToInt(&a))[3] = '\x00';
}

fn bar() void {
    var a: [4]u8 = undefined;
    foo(a);
}
```
does seem to run and work correctly, the immutability of arguments just doesn't let you modify through it without hacks.
I'll try changing the code to use arrays like this soon.

**Comment by tadeokondrak** [comment]
So it works better now, I think. I've tested it on both a struct-array target and a pointer target now.

Though I still think using a function parameter on `@vaArg` is not really usable yet.
`@vaArg` now implicitly uses unary `&` on its argument, and when that's a function argument, that means it's trying to cast away const.
I think then the "correct" way to use `@vaArg` on a function parameter with the current code would be to use `@intToPtr`/`@ptrToInt`, which is not great.

File: test/compile_errors.zig (hunk 1 above)
Comment (LakeByTheWoods) [reply test/compile_errors.zig:7438]: Ahhh, I new there had to be at least *one* regression. Yeah it's a bug. Would be fixed with the fix for #5343 if we want that as well.

**Comment by Rocknest** [comment]
@tadeokondrak why not always define builtin.VaList as `*@Type(.Opaque)`? And `@vaStart` can return `VaListImpl`, defined as:
```zig
// for windows
pub const VaListImpl = struct {
  value: *u8,
  pub fn ref(self: *@This()) VaList {
    return @ptrCast(VaList, self);
    // @vaArg can mutate value @ptrCast(*VaListImpl, ap).*.value = ...
  }
}

// for struct
pub const VaListImpl = extern struct {
  gp_offset: c_uint,
  fp_offset: c_uint,
  overflow_arg_area: *c_void,
  reg_save_area: *c_void,
  pub fn ref(self: *@This()) VaList {
    return @ptrCast(VaList, self);
  }
}
```
`@vaEnd()` implicitly calls `.ref()` on the argument, if it requires a pointer internally.

```zig
fn vadd(n: c_int, ap: builtin.VaList) callconv(.C) c_int {
    var i: c_int = 0;
    var res: c_int = 0;
    while (i < n) : (i += 1) {
        res += @vaArg(ap, c_int);
    }
    return res;
}

fn add(n: c_int, ...) callconv(.C) c_int {
    var ap = @vaStart();
    defer @vaEnd(ap);
    return vadd(n, ap.ref());
}
```

I think uninitialized `va_list` can be created with `var ap: @TypeOf(@vaStart()) = undefined;` without reaching touching VaListImpl

**Comment by andrewrk** [comment]
I'm closing this because it is an old draft. However please feel free to re open a new PR and pick up where you left off.

**Pull request closed without merging**